internal/controllers/token_exchange.go

```diff
@@ -276,59 +278,48 @@ func (t *TokenExchangeController) evaluateSacdDoc(c *fiber.Ctx, record *models.P
 
 	// If any permissions are missing, return an error
 	if len(missingPermissions) > 0 {
-		return fiber.NewError(fiber.StatusBadRequest,
-			fmt.Sprintf("Address %s lacks permissions %v on token id %d for asset %s.",
-				grantee.Hex(), missingPermissions, pr.TokenID, pr.NFTContractAddress))
+		return fmt.Errorf("missing permissions: %v on token id %d for asset %s", missingPermissions, tokenReq.TokenID, tokenReq.NFTContractAddress)
 	}
 
-	// If we get here, all permissions are valid
-	return t.createAndReturnToken(c, pr, grantee)
+	return nil
 }
 
-func intArrayTo2BitArray(indices []int64, length int) (*big.Int, error) {
-	mask := big.NewInt(0)
-
-	for _, index := range indices {
-		if index < 0 || index >= int64(length) {
-			return big.NewInt(0), fmt.Errorf("invalid index %d. These must be non-negative and less than %d", index, length)
+func evaluateCloudEvents(agreement map[string]map[string]*shared.StringSet, tokenReq *TokenRequest) error {
+	var err error
+	for _, req := range tokenReq.CloudEvents.Events {
+		grantedAggs, ok := agreement[req.EventType]
+		if !ok {
+			err = errors.Join(err, fmt.Errorf("lacking grant for requested event type: %s", req.EventType))
+			continue
 		}
-		mask.SetBit(mask, int(index*2), 1)
-		mask.SetBit(mask, int(index*2+1), 1)
-	}
 
-	return mask, nil
-}
+		source := tokenclaims.GlobalAttestationPermission
+		if req.Source != nil {
+			source = *req.Source
+		}
 
-// validateAssetDID verifies that the provided DID matches the NFT contract address
-// and token ID specified in the permission token request.
-//
-// Parameters:
-//   - did: The decentralized identifier string to validate, typically in the format "did:nft:..."
-//   - req: The permission token request containing the NFT contract address and token ID to match against
-//
-// Returns:
-//   - bool: true if the DID is valid and matches the request parameters, false otherwise
-//   - error: An error describing why validation failed, or nil if validation succeeded
-func (t *TokenExchangeController) validateAssetDID(did string, req *TokenRequest) (bool, error) {
-	decodedDID, err := cloudevent.DecodeNFTDID(did)
-	if err != nil {
-		return false, fmt.Errorf("failed to decode DID: %w", err)
-	}
+		idSet, ok := grantedAggs[source]
+		if !ok {
+			err = errors.Join(err, fmt.Errorf("lacking %s grant for requested source: %s", req.EventType, source))
+			continue
+		}
 
-	requestNFTAddr := common.HexToAddress(req.NFTContractAddress)
+		if idSet.Len() == 0 {
```

Comment:
We should treat this the same as source where `*` means all so we are consistent.
```suggestion
		if idSet.Contains(tokenclaims.GlobalAttestationPermission) {
```